### PR TITLE
Add correspondence augmenter

### PR DESCRIPTION
**Pull request by ayushbaid**
Correspondence augmenter fits in between the two-view estimator and the multi-view estimator. It uses the pose estimates from the 2-view estimator and runs a different correspondence generator to augmend correspondences

### CODE REVIEW

**Comment by johnwlambert** [comment]
Any preliminary results if this helps, eg. lightglue + loftr?

File: gtsfm/frontend/two_view_ba_utils.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        )
+        track, _, _ = point3d_initializer.triangulate(track2d)
+        if track is not None:
+            triangulated_indices.append(j)
+            triangulated_tracks.append(track)
+
+    return triangulated_indices, triangulated_tracks
+
+
+def bundle_adjust(
```
Comment (johnwlambert) [inline gtsfm/frontend/two_view_ba_utils.py:59]: Thanks for the PR @ayushbaid. QQ -- don't we already have code for 2-view BA here? https://github.com/borglab/gtsfm/blob/master/gtsfm/two_view_estimator.py#L123

Didn't quite follow the difference with the code here

File: gtsfm/runner/gtsfm_runner_base.py [hunk 2]
```diff
@@ -252,6 +258,20 @@ def run(self) -> GtsfmData:
                 gt_scene_mesh=self.loader.get_gt_scene_trimesh(),
             )
 
+            if self.scene_optimizer.correspondence_augmenter:
```
Comment (johnwlambert) [inline gtsfm/runner/gtsfm_runner_base.py:261]: @ayushbaid would it be cleaner to get the two set of correspondences, and merge their keypoints with the KeypointAggregator, and then run the two-view estimator once? Don't we only want to estimate (R,t) once with all the correspondences?

File: gtsfm/frontend/two_view_ba_utils.py (hunk 1 above)
Comment (johnwlambert) [reply gtsfm/frontend/two_view_ba_utils.py:59]: Oh I think I see now, we are just moving the other code to this new file.

File: gtsfm/frontend/correspondence_augmenter.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        for keypoints_original, keypoints_new in zip(keypoints_list, keypoints_to_augment_list):
+            keypoints_augmented, offset = self._augment_keypoints(
+                keypoints_original=keypoints_original, keypoints_new=keypoints_new
+            )
+            augmented_keypoints_list.append(keypoints_augmented)
+            keypoint_idx_offsets.append(offset)
+
+        # Merge the corr_idxs
+        results: Dict[Tuple[int, int], TWO_VIEW_OUTPUT] = {}
+        for (i1, i2), two_view_output in two_view_outputs.items():
```
Comment (johnwlambert) [inline gtsfm/frontend/correspondence_augmenter.py:138]: Could we share the code block here with the code here? https://github.com/borglab/gtsfm/blob/master/gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_unique.py#L43-L56 Is it identical?

File: gtsfm/frontend/correspondence_augmenter.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        gt_scene_mesh: Optional[Any],
+    ) -> Tuple[List[Keypoints], Dict[Tuple[int, int], TWO_VIEW_OUTPUT]]:
+        (
+            keypoints_to_augment_list,
+            putative_corr_idxs_to_augment,
+        ) = self._correspondence_generator.generate_correspondences(
+            client=client, images=images, image_pairs=list(two_view_outputs.keys())
+        )
+
+        def verify_putative_correspondences(
```
Comment (johnwlambert) [inline gtsfm/frontend/correspondence_augmenter.py:75]: Just curious, have you checked if we get similar accuracy from 

1. Merge unverified, and estimate 2-view w/ verification once.
2. Match and verify separately, then merge? 

My understanding was that https://www.kaggle.com/competitions/image-matching-challenge-2023/discussion/417407 did (1) not (2)
@travisdriver fyi

File: gtsfm/frontend/two_view_ba_utils.py (hunk 1 above)
Comment (ayushbaid) [reply gtsfm/frontend/two_view_ba_utils.py:59]: Yes. It helps me reuse it here. Although I am wondering if I should just do triangulation in correspondence augmenter and skip BA.

File: gtsfm/runner/gtsfm_runner_base.py (hunk 2 above)
Comment (ayushbaid) [reply gtsfm/runner/gtsfm_runner_base.py:261]: I thought about that approach too, but I am wondering if having _lower_ quality correspondences before 2-view estimator will result in poor pose estimates. What do you think?

File: gtsfm/frontend/correspondence_augmenter.py (hunk 4 above)
Comment (ayushbaid) [reply gtsfm/frontend/correspondence_augmenter.py:75]: Yes I saw that too. Let me work on approach 1 too then.

Comment (travisdriver) [reply gtsfm/frontend/correspondence_augmenter.py:75]: Yes, I think (1) makes more sense

**Comment by ayushbaid** [comment]
Closing as #663 seems like a simple and preferred approach.

**Pull request closed without merging**